internal/controller/instance_controller.go

```diff
@@ -0,0 +1,1628 @@
+	"go.datum.net/infra-provider-gcp/internal/downstreamclient"
+	datumhandler "go.datum.net/infra-provider-gcp/internal/handler"
+	"go.datum.net/infra-provider-gcp/internal/locationutil"
+	datumsource "go.datum.net/infra-provider-gcp/internal/source"
+	networkingv1alpha "go.datum.net/network-services-operator/api/v1alpha"
+	computev1alpha "go.datum.net/workload-operator/api/v1alpha"
+)
+
+const crossplaneFinalizer = "finalizer.managedresource.crossplane.io"
+
+var errResourceIsDeleting = errors.New("resource is deleting")
+
+var imageMap = map[string]string{
+	"datumcloud/ubuntu-2204-lts":           "projects/ubuntu-os-cloud/global/images/ubuntu-2204-jammy-v20240927",
+	"datumcloud/cos-stable-117-18613-0-79": "projects/cos-cloud/global/images/cos-stable-117-18613-0-79",
+}
+
+var machineTypeMap = map[string]string{
+	"datumcloud/d1-standard-2": "n2-standard-2",
+}
+
+//go:embed cloudinit/populate_secrets.py
+var populateSecretsScript string
+
+// InstanceReconciler reconciles Instances and manages their intended state in
+// GCP
+type InstanceReconciler struct {
+	mgr               mcmanager.Manager
+	Config            config.GCPProvider
+	LocationClassName string
+	DownstreamCluster cluster.Cluster
+}
+
+func (r *InstanceReconciler) Reconcile(ctx context.Context, req mcreconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	cl, err := r.mgr.GetCluster(ctx, req.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	ctx = mccontext.WithCluster(ctx, req.ClusterName)
+
+	var instance computev1alpha.Instance
+	if err := cl.GetClient().Get(ctx, req.NamespacedName, &instance); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Don't do anything if a location isn't set
+	if instance.Spec.Location == nil {
+		return ctrl.Result{}, nil
+	}
+
+	location, shouldProcess, err := locationutil.GetLocation(ctx, cl.GetClient(), *instance.Spec.Location, r.LocationClassName)
+	if err != nil {
+		return ctrl.Result{}, err
+	} else if !shouldProcess {
+		return ctrl.Result{}, nil
+	}
+
+	gcpProject := location.Spec.Provider.GCP.ProjectID
+
+	logger.Info("reconciling instance")
+	defer logger.Info("reconcile complete")
+
+	if !instance.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(&instance, gcpInfraFinalizer) {
+			return ctrl.Result{}, r.Finalize(ctx, cl.GetClient(), &instance)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(&instance, gcpInfraFinalizer) {
+		controllerutil.AddFinalizer(&instance, gcpInfraFinalizer)
+		if err := cl.GetClient().Update(ctx, &instance); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if len(instance.Spec.Controller.SchedulingGates) > 0 {
+		logger.Info("instance has scheduling gates, waiting until they are removed", "scheduling_gates", instance.Spec.Controller.SchedulingGates)
+		return ctrl.Result{}, nil
+	}
```

Comment:
Not something we need to tackle as part of this PR, but this feels like good information that could be surfaced through the event recorder so users could see this when doing a describe on the instance resource.

Reply:
> I do agree that emitting events would be useful, but we need to think through how we handle the storage of those. Typically they're not kept for all that long, so the value of events diminishes quickly if you're not actively looking for them.

Agree here.